File: lang/java/SseHelper.java
package ai.axme.examples;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.*;

/**
 * Shared SSE helper for agent intent stream polling.
 * The Java SDK does not include a built-in SSE/listen method,
 * so this helper provides a minimal SSE client for the agent inbox stream.
 */
public final class SseHelper {
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private SseHelper() {}

    /**
     * Poll the agent intent stream (SSE endpoint).
     * Returns parsed intent deliveries from the stream.
     */
    public static List<Map<String, Object>> pollAgentStream(
            String baseUrl, String apiKey, String agentAddress,
            int since, int waitSeconds) throws Exception {

        String path = agentAddress.startsWith("agent://")
                ? agentAddress.substring("agent://".length()) : agentAddress;
        String url = baseUrl + "/v1/agents/" + path + "/intents/stream"
                + "?since=" + since + "&wait_seconds=" + waitSeconds;

        HttpURLConnection conn = (HttpURLConnection) new URL(url).openConnection();
        conn.setRequestProperty("x-api-key", apiKey);
        conn.setRequestProperty("Accept", "text/event-stream");
        conn.setConnectTimeout(5000);
        conn.setReadTimeout((waitSeconds + 5) * 1000);

        List<Map<String, Object>> results = new ArrayList<>();
        try (BufferedReader reader = new BufferedReader(
                new InputStreamReader(conn.getInputStream()))) {
            String eventType = "";
            StringBuilder data = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                if (line.startsWith("event: ")) {
                    eventType = line.substring(7).trim();
                } else if (line.startsWith("data: ")) {
                    data.append(line.substring(6));
                } else if (line.isEmpty() && data.length() > 0) {
                    if (eventType.startsWith("intent.")) {
                        results.add(parseJson(data.toString()));
                    }
                    eventType = "";
                    data.setLength(0);
                }
            }
        }
        return results;
    }

    @SuppressWarnings("unchecked")
    public static Map<String, Object> parseJson(String json) {
        try {
            return MAPPER.readValue(json, Map.class);
        } catch (Exception e) {
            return Map.of();
        }
    }

    public static String env(String key, String def) {
        String v = System.getenv(key);
        return v != null && !v.isEmpty() ? v : def;
    }

    public static String str(Map<String, Object> m, String key) {
        return str(m, key, "");
    }

    public static String str(Map<String, Object> m, String key, String def) {
        Object v = m.get(key);
        return v instanceof String s && !s.isEmpty() ? s : def;
    }

    /** Extract effective payload (handles step-intent wrapping). */
    @SuppressWarnings("unchecked")
    public static Map<String, Object> effectivePayload(Map<String, Object> intent) {
        Map<String, Object> raw = intent.containsKey("payload")
                ? (Map<String, Object>) intent.get("payload") : Map.of();
        if (raw == null) raw = Map.of();
        Object pp = raw.get("parent_payload");
        return pp instanceof Map ? (Map<String, Object>) pp : raw;
    }

    /** Unwrap intent from response. */
    @SuppressWarnings("unchecked")
    public static Map<String, Object> unwrapIntent(Map<String, Object> resp) {
        Object i = resp.get("intent");
        return i instanceof Map ? (Map<String, Object>) i : resp;
    }

    /** Check if intent status is actionable. */
    public static boolean isActionable(Map<String, Object> intent) {
        String status = str(intent, "lifecycle_status");
        if (status.isEmpty()) status = str(intent, "status");
        return Set.of("CREATED", "DELIVERED", "ACKNOWLEDGED", "IN_PROGRESS", "WAITING")
                .contains(status.toUpperCase());
    }

    /** Create RequestOptions with owner_agent set. */
    public static dev.axme.sdk.RequestOptions withOwner(String ownerAgent) {
        return new dev.axme.sdk.RequestOptions(null, null, ownerAgent, null, null);
    }
}

File: lang/dotnet/SseHelper.cs
using System.Net.Http.Headers;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace AxmeExamples;

/// <summary>
/// Shared SSE helper for agent intent stream polling.
/// The .NET SDK does not include a built-in SSE/listen method.
/// </summary>
public static class SseHelper
{
    private static readonly HttpClient Http = new();

    public static async Task<List<JsonObject>> PollAgentStreamAsync(
        string baseUrl, string apiKey, string agentAddress,
        int since = 0, int waitSeconds = 15, CancellationToken ct = default)
    {
        var path = agentAddress.StartsWith("agent://") ? agentAddress["agent://".Length..] : agentAddress;
        var url = $"{baseUrl}/v1/agents/{path}/intents/stream?since={since}&wait_seconds={waitSeconds}";

        using var req = new HttpRequestMessage(HttpMethod.Get, url);
        req.Headers.Add("x-api-key", apiKey);
        req.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("text/event-stream"));

        using var resp = await Http.SendAsync(req, HttpCompletionOption.ResponseHeadersRead, ct);
        resp.EnsureSuccessStatusCode();

        var results = new List<JsonObject>();
        using var reader = new StreamReader(await resp.Content.ReadAsStreamAsync(ct));

        var eventType = "";
        var data = "";

        while (await reader.ReadLineAsync(ct) is { } line)
        {
            if (line.StartsWith("event: ")) eventType = line[7..].Trim();
            else if (line.StartsWith("data: ")) data += line[6..];
            else if (line.Length == 0 && data.Length > 0)
            {
                if (eventType.StartsWith("intent."))
                {
                    var obj = JsonNode.Parse(data)?.AsObject();
                    if (obj != null) results.Add(obj);
                }
                eventType = "";
                data = "";
            }
        }
        return results;
    }

    public static string Env(string key, string def = "") =>
        Environment.GetEnvironmentVariable(key) is { Length: > 0 } v ? v : def;

    public static string Str(JsonObject obj, string key, string def = "") =>
        obj.TryGetPropertyValue(key, out var v) && v is JsonValue jv ? jv.ToString() : def;

    public static JsonObject EffectivePayload(JsonObject intent)
    {
        var raw = intent["payload"]?.AsObject() ?? new JsonObject();
        return raw["parent_payload"]?.AsObject() ?? raw;
    }

    public static JsonObject UnwrapIntent(JsonObject resp) =>
        resp["intent"]?.AsObject() ?? resp;

    public static bool IsActionable(JsonObject intent)
    {
        var status = (Str(intent, "lifecycle_status") is { Length: > 0 } s ? s : Str(intent, "status")).ToUpper();
        return status is "CREATED" or "DELIVERED" or "ACKNOWLEDGED" or "IN_PROGRESS" or "WAITING";
    }
}
